llm/data_gemma/base.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Base Types."""
+
+import dataclasses
+from typing import Any, Protocol
+
+
+DC = '__DC__'
+
+
+@dataclasses.dataclass(frozen=True)
+class Options:
+  """Common options for all APIs."""
+  # Print messages to stdout.
+  verbose: bool = True
+
+  def vlog(self, msg: str) -> None:
+    if self.verbose:
+      print(msg)
+
+
+@dataclasses.dataclass(frozen=True)
+class LLMCall:
+  prompt: str
+  response: str
+  duration_secs: float
+  error: str | None = None
+
+  def debug(self, i: int = 0) -> str:
+    return (
+        f'\n### Prompt {i} ###\n{self.prompt}\n'
+        f'### Response {i} ###\n{self.response}\n'
+        f'### LLM Duration {i} {self.duration_secs}s ###\n'
+    )
+
+
+@dataclasses.dataclass
+class DCCall:
+  """A single response from Data Commons."""
+
+  id: int = 0
+  query: str = ''
+
+  # For point: val and date is set
+  val: str = ''
+  date: str = ''
+  # For table: table is set
+  table: str = ''
+
+  unit: str = ''
+  title: str = ''
+  src: str = ''
+  url: str = ''
+  var: str = ''
+  score: float = 0.0
+
+  # The original LLM Value in case of RIG.
+  llm_val: str = ''
+
+  def footnote(self) -> str:
+    return (
+        f'Per {self.src}, value was {self.val}{self._dunit()} in {self.date}.'
+        f' See more at {self.url}'
+    )
+
+  def debug(self) -> str:
+    if not self.title:
+      return ''
+    if self.table:
+      return self.answer()
+    return (
+        f'"{self.title}" was {self.val}{self._dunit()} in'
+        f' {self.date} per {self.src} ({self.var}:{self.score})'
+    )
+
+  def answer(self) -> str:
+    if self.table:
+      return f'{self.header()}\n{self.table}'
+    else:
+      return (
+          f'According to {self.src}, "{self.title}" was'
+          f' {self.val}{self._dunit()} in {self.date}.'
+      )
+
+  def header(self) -> str:
+    if self.table:
+      if self.unit:
+        header = f'{self.title} (unit: {self.unit})'
+      else:
+        header = f'{self.title}'
+      return f'{header}, according to {self.src}'
+
+    return self.title
+
+  def val_and_unit(self) -> str:
+    return f'{self.val}{self._dunit()}'
+
+  def _dunit(self) -> str:
+    return ' ' + self.unit if self.unit else ''
+
+
+@dataclasses.dataclass(frozen=True)
```

Comment:
nit: move this after Flow definition

Reply:
Flow definition needs FlowResponse, so I think this should come before